lib/concurrency/Prefix_Refinement.thy

```diff
@@ -938,163 +1189,614 @@ theorem prefix_refinement_bind_general[rule_format]:
+  "\<lbrakk>prefix_refinement sr isr intsr (f \<oplus> dc) AR R P Q a c;
+    prefix_refinement sr intsr osr (f \<oplus> rvr) AR R P'' Q'' b d;
+    \<lbrace>P'\<rbrace>, \<lbrace>AR\<rbrace> a -, \<lbrace>\<lambda>_. P''\<rbrace>, \<lbrace>E\<rbrace>; \<lbrace>Q'\<rbrace>, \<lbrace>R\<rbrace> c -, \<lbrace>\<lambda>_. Q''\<rbrace>, \<lbrace>E'\<rbrace>;
+    \<And>rv s0 s rv' t0 t. \<lbrakk>E rv s0 s; E' rv' t0 t; intsr s t\<rbrakk> \<Longrightarrow> osr s t\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr (f \<oplus> rvr) AR R (P and P') (Q and Q') (a >>=E (\<lambda>_. b)) (c >>=E (\<lambda>_. d))"
+  by (rule prefix_refinement_bindE; assumption)
+
+lemmas prefix_refinement_bind_mapr = prefix_refinement_bind[where rvr'="(=) \<circ> g" for g, simplified]
+lemmas prefix_refinement_bind_maprE = prefix_refinement_bindE[where rvr'="(=) \<circ> g" for g, simplified]
+
+
+section \<open>Rules for walking prefix refinement into basic constructs\<close>
+
+lemma prefix_refinement_if:
+  "\<lbrakk>G = G'; prefix_refinement sr isr osr rvr AR R P Q a c;
+    prefix_refinement sr isr osr rvr AR R P' Q' b d \<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr rvr AR R
+               (if G then P else P') (if G' then Q else Q')
+               (if G then a else b) (if G' then c else d)"
+  by simp
+
+\<comment> \<open>FIXME: copied from Word_Lib.Many_More, where would be a good spot to put it?\<close>
+lemma if_apply_def2:
+  "(if P then F else G) = (\<lambda>x y. (P \<longrightarrow> F x y) \<and> (\<not> P \<longrightarrow> G x y))"
+  by simp
+
+\<comment> \<open>FIXME: this could have slightly better bound variable names if written out, should we just do
+           that and avoid the previous FIXME?\<close>
+lemmas prefix_refinement_if2 = prefix_refinement_if[unfolded if_apply_def2]
+
+lemma prefix_refinement_when:
+  "\<lbrakk>G = G'; prefix_refinement sr isr isr rvr AR R P Q a c; rvr () ()\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr isr rvr AR R (\<lambda>x y. G \<longrightarrow> P x y) (\<lambda>x y. G' \<longrightarrow> Q x y)
+         (when G a) (when G' c)"
+  unfolding when_def
+  apply clarsimp
+  apply (rule prefix_refinement_return_imp)
+  apply simp
   done
 
-abbreviation(input)
+lemma prefix_refinement_whenE:
+  "\<lbrakk>G = G'; prefix_refinement sr isr isr (f \<oplus> rvr) AR R P Q a c; rvr () ()\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr isr (f \<oplus> rvr) AR R (\<lambda>x y. G \<longrightarrow> P x y) (\<lambda>x y. G' \<longrightarrow> Q x y)
+         (whenE G a) (whenE G' c)"
+  unfolding whenE_def returnOk_def
+  apply clarsimp
+  apply (rule prefix_refinement_return_imp)
+  apply simp
+  done
+
+\<comment> \<open>FIXME: an equivalent doesn't seem to exist for corres, do we want it to or does it mean we
+          shouldn't bother with this?\<close>
+lemma prefix_refinement_unless:
+  "\<lbrakk>G = G'; prefix_refinement sr isr isr rvr AR R P Q a c; rvr () ()\<rbrakk>
```

Comment:
That would be really good to add to corres. We always unfold `unless_def` for no good reason.